src/eredis_cluster_monitor.erl

```diff
@@ -164,45 +164,132 @@ get_cluster_slots(InitNodes, Options) ->
 
 %% =============================================================================
-%% @doc Generic function to do cluster information requests to a init node
+%% @doc Fetch cluster information from an already connected node or from an init
+%%      node if no nodes are connected. Throws an exception if everything fails.
 %% @end
 %% =============================================================================
-get_cluster_info([], _Options, _Query, _FailFn, ErrorList) ->
+
+-spec get_cluster_info(State :: #state{}, Options :: options(), Query :: list(),
+                       FailFn :: fun((#node{}) -> iodata())) ->
+          iodata().
+get_cluster_info(State, Options, Query, FailFn) ->
+    %% First try to use existing connections in pools found in the slot map
+    case get_cluster_info_from_existing_pools(State#state.slots_maps, Options,
+                                              Query, FailFn) of
+        {ok, Result} ->
+            Result;
+        _Error ->
+            %% No usable pool connected. Connect to init nodes.
+            get_cluster_info_from_init_nodes(State#state.init_nodes, Options,
+                                             Query, FailFn, [])
+    end.
+
+-spec get_cluster_info_from_existing_pools(SlotMaps :: tuple(),
+                                           Options  :: options(),
+                                           Query    :: list(),
+                                           FailFn   :: fun((#node{}) -> list())) ->
+          {ok, list()} | {error, no_connection}.
+get_cluster_info_from_existing_pools(SlotMaps, Options, Query, FailFn) ->
+    get_cluster_info_from_existing_pools(SlotMaps, Options, Query, FailFn, {1, #{}}).
+
+get_cluster_info_from_existing_pools(SlotMaps, Options, Query, FailFn, SlotMapIterator) ->
+    case next_node_in_slots_maps(SlotMaps, Options, SlotMapIterator) of
+        {ok, Node, NewSlotMapIterator} ->
+            Transaction = fun(Connection) ->
+                                  get_cluster_info_from_connetion(Connection, Query, FailFn, Node)
+                          end,
+            try
+                {ok, _Result} = poolboy:transaction(Node#node.pool, Transaction)
+            catch
+                _:_ ->
+                    get_cluster_info_from_existing_pools(SlotMaps, Options, Query, FailFn, NewSlotMapIterator)
+            end;
+        none ->
+            {error, no_connection}
+    end.
+
+%% Get next node from a tuple of slots maps, using an iterator tuple to remember what's next
+-spec next_node_in_slots_maps(SlotsMaps :: tuple(),
+                              Options   :: options(),
+                              Iterator  :: {Index           :: pos_integer(),
+                                            ExcludePoolsMap :: #{atom() => true}}) ->
+          {ok, Node :: #node{}, NextIterator :: {NextIndex          :: pos_integer(),
+                                                 NewExcludePoolsMap :: #{atom() => true}}} |
```

Comment:
is the ExcludePoolsMap to avoid duplicate attempts on the same redis instance?
Maybe its an FailingPoolsMap or an AttemptedPoolsMap

Reply:
Yes, it's to avoid trying the same pool twice.

From the perspective of this function, there is no concept of anything failing. Only that it has been returned already and therefore it shouldn't be returned again. Does this make sense?

Reply:
On second though, I like AttemptedPoolsMap. :-)